### PR TITLE
Updated memory management, added unit tests, fixed SonarCloud issues

### PR DESCRIPTION
**Pull request by uweseimet**
This PR resolves most of the remaining memory management and structural issues. The vast majority of SonarCloud issues is now caused by the horrible code in filesystem.cpp/h.
Testing was done with my standard hardware setup on buster and bullseye, with gcc and clang, and of course with the extended set of unit tests.

@akuker I would also like to merge this PR myself after approval, in order to provide details in the merge message.

### CODE REVIEW

File: src/raspberrypi/devices/scsimo.cpp [hunk 1]
```diff
@@ -21,11 +19,18 @@
...
 
-SCSIMO::SCSIMO(const unordered_set<uint32_t>& sector_sizes, const unordered_map<uint64_t, Geometry>& geometries)
-	: Disk("SCMO")
+SCSIMO::SCSIMO(const unordered_set<uint32_t>& sector_sizes) : Disk("SCMO")
 {
 	SetSectorSizes(sector_sizes);
-	SetGeometries(geometries);
+
+	// 128 MB, 512 bytes per sector, 248826 sectors
+	geometries[0x797f400] = make_pair(512, 248826);
```
Comment (akuker) [inline src/raspberrypi/devices/scsimo.cpp:27]: @uweseimet - Would you have any objections with changing these hex values so we can see where they come from? If I understand them correctly,

geometries[(512*248826)] = make_pair(512,248826);

File: src/raspberrypi/devices/scsimo.h [hunk 2]
```diff
@@ -5,12 +5,10 @@
 //
 //	Copyright (C) 2001-2006 ＰＩ．(ytanaka@ipc-tokai.or.jp)
 //	Copyright (C) 2014-2020 GIMONS
-//  	Copyright (C) akuker
+//  Copyright (C) akuker
```
Comment (akuker) [inline src/raspberrypi/devices/scsimo.h:8]: some spacing issues here.

**Review by akuker — CHANGES_REQUESTED** [review]

**Comment by akuker** [comment]
Thanks @uweseimet !! Just a couple of comments.

I know I'm slow, but could you avoid changes to gpiobus for the time being? (unless they're bug fixes)

File: src/raspberrypi/devices/scsimo.cpp (hunk 1 above)
Comment (uweseimet) [reply src/raspberrypi/devices/scsimo.cpp]: Yes, that's definitely better. (The hex values still came from the very old code.) Done.

File: src/raspberrypi/devices/scsimo.h (hunk 2 above)
Comment (uweseimet) [reply src/raspberrypi/devices/scsimo.h]: I have not changed anything here, but just applied the change anyway :).

**Comment by uweseimet** [comment]
@akuker Regarding the gpiobus changes, these fix SonarCloud issues. Just a few lines are changed. If this is an issue for you we can revert the changes in the 'hal' folder to what we currently have in develop. On the other hand, except for a few more changes lines (currently in the feature_more_unit_tests branch) I am done with updating 'hal'. I just merged these pending changes into the PR for feature_memory_management.
So we have two alternatives now:

1. Keep everything in 'hal' as it is right now, with no changes planned from my side.
2. Revert 'hal' to what's currently in develop. In this case I would like to ask you for applying the pending changes as part of your upcoming changes.

I hope it's clear what I mean :).

**Comment by akuker** [comment]
> I hope it's clear what I mean :). Of course I prefer option 1, because these changes jsut a few lines and because they have already been tested by @rdmark and me.

Option 1 is good for me :-) The changes should be easy enough to merge in to my stuff. Have a great weekend!

**Review by akuker — APPROVED** [review]

**Comment by uweseimet** [comment]
@akuker Once more, thank you for taking time to review my numerous changes. Have a great weekend as well!

**Pull request merged**